src/rand/weighted_shuffle.zig

```diff
@@ -0,0 +1,418 @@
+                index = (index << BIT_SHIFT) + FANOUT;
+            }
+            return .{ index - self.tree.items.len, weight };
+        }
+
+        pub fn removeIndex(self: *WeightedShuffle(T), k: usize) void {
+            var index = self.tree.items.len + k;
+            var weight: T = 0;
+            while (index != 0) {
+                const offset = index & BIT_MASK;
+                index = (index - 1) >> BIT_SHIFT;
+                if (offset > 0) {
+                    if (self.tree.items[index][offset - 1] != weight) {
+                        self.remove(k, self.tree.items[index][offset - 1] - weight);
+                    } else {
+                        self.removeZero(k);
+                    }
+                    return;
+                }
+                for (self.tree.items[index]) |node| {
+                    weight += node;
+                }
+            }
+            if (self.weight != weight) {
+                self.remove(k, self.weight - weight);
+            } else {
+                self.removeZero(k);
+            }
+        }
+
+        pub fn removeZero(self: *WeightedShuffle(T), k: usize) void {
+            // TODO: Replace with better implementation.
+            var found = false;
+            for (self.zeros.items, 0..) |i, j| {
+                if (i == k) {
+                    found = true;
+                } else if (found) {
+                    self.zeros.items[j - 1] = j;
+                }
+            }
+            if (found) _ = self.zeros.pop();
+        }
+
+        /// Returns the index of the first sampled weight.
+        /// The function is non-destructive and does not remove the weights from the internal state.
+        pub fn first(self: *const WeightedShuffle(T), rng: Random) ?usize {
+            if (self.weight > 0) {
+                const sample = uintLessThan(T, rng, self.weight);
+                const index, _ = self.search(sample);
+                return index;
+            }
+            if (self.zeros.items.len == 0) {
+                return null;
+            }
+            const index = uintLessThan(usize, rng, self.zeros.items.len);
+            return self.zeros.items[index];
+        }
+
+        /// Returns a shuffled list of weights.
+        /// The function is destructive and removes the weights from the internal state.
+        pub fn shuffle(self: *WeightedShuffle(T), rng: Random) Iterator {
+            return .{
+                .shuffle = self,
+                .rng = rng,
+                .index = 0,
+            };
+        }
+
+        fn getTreeSize(count: usize) usize {
+            var size: usize = if (count == 1) 1 else 0;
+            var nodes: usize = 1;
+            while (nodes < count) {
+                size += nodes;
+                nodes *= FANOUT;
+            }
+            return size;
+        }
+
+        /// Custom Rng downsampling function designed to match the agave sampler used
```

Comment:
Is it actually necessary to reproduce the exact same calculation as agave? For something like the leader schedule, it is strictly necessary because every node needs to deterministically calculate the same exact schedule. But for turbine, I did not realize that determinism is required. As long as the results are statistically similar, isn't that sufficient?

Reply:
Yeah I think you are correct, however it wasn't too much work and I found it useful for confirming that the actual implementation was correct. Its also used elsewhere in the validator so I figured it would just be easier to implement the same downsampler than inspect other use cases to understand if we need determinism

Reply:
For posterity... I misunderstood turbine here. It *does* need to be identical. Every node needs to calculate the entire turbine tree. Every node is expected to agree on the contents of the entire tree.

My incorrect assumption was that each node would select their own downstream nodes. The tree would be built through a series of stochastic decisions made by each node. In that case, it would be a graph that only approximately resembles a tree. This is similar to how gossip works, but it's not how turbine works.